Onchain permissioning (#1300)
PegaSysEng/pantheon: docs/Permissions/Onchain-Permissioning.md

description: Onchain Permissioning
<!--- END of page meta data -->

# Onchain Permissioning 

Onchain permissioning uses smart contracts to store and maintain the node whitelist. Using onchain permissioning
enables all nodes to read the node whitelist from one location. 
                                                                       
!!! note
    Pantheon supports onchain permissioning for nodes. Onchain permissioning for accounts will be available 
    in a future Pantheon release. 

Onchain permissioning uses the following smart contracts: 

* Ingress - a simple contract functioning as a gateway to the Rules contract. The Ingress contract is deployed 
to a static address. 

> * Rules - stores the node whitelist and list of admin accounts. 

Reviewer: the Rules and Admin can be (and are likely to be very soon) separate contracts - don't know if you want to be more specific here

Reviewer: Rules - stores the node whitelist.
Admin - stores the list of admin accounts.

Author: Fixed

## Pre-requisites 

> For nodes that are going to set up or maintain the nodes whitelist: 

Reviewer: and/or the list of admins

Author: Fixed

* [NodeJS](https://nodejs.org/en/) v8.9.4 or later 

* [Truffle](https://truffleframework.com/docs/truffle/getting-started/installation)

## Add Ingress Contract to Genesis File

Add the Ingress contract to the genesis file for your network by copying it from [`genesis.json`](https://github.com/PegaSysEng/permissioning-smart-contracts/blob/master/genesis.json) 
in the [`permissioning-smart-contracts` repository](https://github.com/PegaSysEng/permissioning-smart-contracts): 

>    

Reviewer: do we want to explain how to get this code?

Author: Yep what Nico said

```json
"0x0000000000000000000000000000000000009999": {
      "comment": "Ingress smart contract",
      "balance": "0",
      "code": <stripped>,
      "storage": {
         <stripped>
      }
}
```

!!! important 

>     Ensure your genesis file includes a sufficient milestone to support the permissioning contracts. 

Reviewer 2: I think we definitely need to specify a minimum milestone here.

>     Ensure your genesis file includes a sufficient milestone to support the permissioning contracts. 

Author: Fixed

    * [--permissions-nodes-contract-enabled](../Reference/Pantheon-CLI-Syntax.md#permissions-nodes-contract-enabled)
      to enable onhcain permissioning

    * [--permissions-nodes-contract-address](../Reference/Pantheon-CLI-Syntax.md#permissions-nodes-contract-address)
      set to the address of the Ingress contract in the genesis file (`"0x0000000000000000000000000000000000009999"`)

1. Create the following environment variables and set to the specified values: 

    * `PANTHEON_NODE_PERM_ACCOUNT` - address of account used to interact with the permissioning contracts. 

    * `PANTHEON_NODE_PERM_KEY` - private key of the account used to interact with the permissioning contracts.

    * `INGRESS_CONTRACT_ADDRESS` - address of the Ingress contract in the genesis file.  

    * `PANTHEON_NODE_PERM_ENDPOINT` - required only if your node is not using the default JSON-RPC host and port (`http://127.0.0.1:8545`). 
       Set to JSON-RPC host and port. 

    !!! note
        If your network is not configured as a [free gas network](../Configuring-Pantheon/FreeGas.md), the account used to 
        interact with the permissioning contracts must have a balance. 

> 1. Clone the `permissioning-smart-contracts` repository: 

Reviewer 3: Think we also need to specify that it needs to be a mining node

> 1. Clone the `permissioning-smart-contracts` repository: 

Author: Done

> 1. Clone the `permissioning-smart-contracts` repository: 

Author: @NicolasMassart - that's going to have knockon effects to later and I think the whole thing could probably benefit from a rework of the flow but that'll have to be post 1.1.  I'll raise a JI for this and include your comment.

    ```bash 
    git clone https://github.com/PegaSysEng/permissioning-smart-contracts.git
    ```

1. Change into the `permissioning-smart-contracts` directory and run: 

    ```bash
    npm install
    ```

> ## Deploy Rules Contract 

Reviewer: ```suggestion
## Deploy Rules and Admin Contracts 
```

Author: Fixed

> 

Reviewer: make it clearer that only one node does this action. Ingress is shared at a specific location because it's in the genesis file, and the Rules contract is deployed (and registered with the Ingress contract) using truffle migrate. Only one node does this (there's nothing stopping you doing it a second time but then each truffle console will be talking to a different Rules contract).

Reviewer: Actually, what will stop you from doing it a second time is - whether the account you're using is an admin account.

Author: Added an 'important' to make this clearer

> In the `permissioning-smart-contracts` directory, deploy the Rules contract: 

Reviewer: ```suggestion
In the `permissioning-smart-contracts` directory, deploy the Rules and Admin contracts: 
```

Author: Fixed

```bash
truffle migrate --reset
```

> The Rules contract is deployed and the Ingress contract updated with the name and version of the Rules contract. 

Reviewer: The migration logs the address of the Rules contract.

Author: Fixed

!!! important 
    The account that deploys the Rules contract is automatically an [Admin account](#add-and-remove-admin-accounts). 

## Add and Remove Nodes from the Whitelist 

!!! note  
    Only [Admin accounts](#add-and-remove-admin-accounts) can add or remove nodes from the whitelist. 

To add or remove nodes: 

1. In the `permissioning-smart-contracts` directory, run the Truffle console: 

    ```bash
    truffle console 
    ```

1. Open https://permissioning-tools.pegasys.tech/

1. Enter the [enode URL](../Configuring-Pantheon/Node-Keys.md#enode-url) of the node to be added or removed

> and click the *Process* button. 

Reviewer: ```suggestion
```

Author: Fixed

> 1. Click the *Copy to clipboard* button for _Add Enode_ or _Remove Enode_. 

Reviewer: ```suggestion
1. Click the *Add Node* or *Remove Node* button to display the Truffle commands, and then click *Copy to clipboard*  
```

Author: Fixed

1. Paste the copied command into the Truffle Console. 

   When the transaction is included in a block, the transaction receipt is displayed. 

!!! important 

>     When starting the network, the first node must add itself to the whitelist.  

Author: Yep using this command - have moved this point to the start of the section and updated to indicate this must be done after deploying the admin and rules contracts.

Added nodes do not attempt to reconnect until peer discovery restarts.  To add a whitelisted node as a peer

> without waiting for peer discovery to restart, use [`admin_addPeer`](../Reference/JSON-RPC-API-Methods.md#admin_addpeer). 

Reviewer: if we are insisting that all added nodes have node 1 as bootnode, they should not need to call addPeer because we connect to bootnodes on startup. (Once we do the staticnodes change this may be different.)

Author: If you've already got the node running before adding it to the whitelist, then you do need to call admin_addPeer.  I'll make this clearer.

## Display Nodes Whitelist 
 
To display the nodes whitelist, paste the following into the Truffle Console: 

> ```bash

Reviewer 4: The syntax highlight for Truffle console is more javascript than bash
```suggestion
```javascript
```

Author: Fixed

> Rules.deployed().then(function(instance) {instance.getKeyCount().then(function(txCount) {console.log("size of whitelist: " + txCount); var i=txCount; while(i>=0) {instance.getWhitelistKey(i--).then(function(tx) {console.log(tx)})}});});

Reviewer: ```suggestion
Rules.deployed().then(function(instance) {instance.getSize().then(function(txCount) {console.log("size of whitelist: " + txCount); var i=txCount; while(i>=0) {instance.getByIndex(i--).then(function(tx) {console.log(tx)})}});});
```

Author: Fixed

```

## Start Other Network Nodes 

For participating nodes that are not going to add or remove nodes from the whitelist:

1. Start the nodes with permissioning contract options: 

    * [--permissions-nodes-contract-enabled](../Reference/Pantheon-CLI-Syntax.md#permissions-nodes-contract-enabled)

>      to enable onhcain permissioning

Author: Fixed

>     * [--permissions-nodes-contract-address](../Reference/Pantheon-CLI-Syntax.md#permissions-nodes-contract-address)

Reviewer 4: Don't we have a default value for this if we alway advertise using `0x0000000000000000000000000000000000009999` ?

     set to the address of the Ingress contract in the genesis file (`"0x0000000000000000000000000000000000009999"`)

> 

Reviewer: ```suggestion
    * [--bootnodes](the first node, the node that deployed the Rules and Admin contracts.  )

```

Author: Fixed

1. Copy the [enode URL](../Configuring-Pantheon/Node-Keys.md#enode-url) and [have node added to the whitelist](#add-and-remove-nodes-from-the-whitelist).

For participating nodes that are going to add or remove nodes from the whitelist: 

1. Ensure [prerequisites installed](#pre-requisites).

1. Complete [permissioning setup](#onchain-permissioning-setup).

1. Copy the [enode URL](../Configuring-Pantheon/Node-Keys.md#enode-url) and [have node added to the whitelist](#add-and-remove-nodes-from-the-whitelist).

1. Have account for node that will interact with permissioning contracts added as an [admin account](#add-and-remove-admin-accounts). 

## Add and Remove Admin Accounts      

The account that deploys the Rules contract is automatically an Admin account. Only Admin accounts can
add or remove nodes from the whitelist.

To add Admin accounts, paste the following commands into Truffle Console:

> ```bash tab="Truffle Console Command"

Reviewer 4: The syntax highlight for Truffle console is more javascript than bash
```suggestion
```javascript tab="Truffle Console Command"
```

> Rules.deployed().then(function(instance) {instance.addAdmin("<Admin Account>").then(function(tx) {console.log(tx)});});

Reviewer: Not sure if this will work if you're running this on a truffle console that did not deploy the Rules contract. You might need to use `Rules.at("<rules-address>").then(function(instance) {instance.addAdmin("<Admin Account>").then(function(tx) {console.log(tx)});});`

Author: Fixed

> Rules.deployed().then(function(instance) {instance.addAdmin("<Admin Account>").then(function(tx) {console.log(tx)});});

Reviewer: ```suggestion
Admin.deployed().then(function(instance) {instance.addAdmin("<Admin Account>").then(function(tx) {console.log(tx)});});
```

Author: Fixed

```

> ```bash tab="Example"

Reviewer 4: The syntax highlight for Truffle console is more javascript than bash
```suggestion
```javascript tab="Example"
```

Author: Fixed

Rules.deployed().then(function(instance) {instance.addAdmin("0x627306090abaB3A6e1400e9345bC60c78a8BEf57").then(function(tx) {console.log(tx)});});
```

To remove Admin accounts, paste the following commands into Truffle Console:

> ```bash tab="Truffle Console Command"

Reviewer 4: The syntax highlight for Truffle console is more javascript than bash
```suggestion
```javascript tab="Truffle Console Command"
```

Author: Fixed

Rules.deployed().then(function(instance) {instance.removeAdmin("<Admin Account>").then(function(tx) {console.log(tx)});});
```

> ```bash tab="Example"

Reviewer 4: The syntax highlight for Truffle console is more javascript than bash
```suggestion
```javascript tab="Example"
```

Author: Fixed

Rules.deployed().then(function(instance) {instance.removeAdmin("0x627306090abaB3A6e1400e9345bC60c78a8BEf57").then(function(tx) {console.log(tx)});});
```

### Display Admin Accounts 

To display the list of admin accounts, paste the following into the Truffle Console: 

> ```bash

Reviewer 4: The syntax highlight for Truffle console is more javascript than bash
```suggestion
```javascript
```

Author: Fixed

> Rules.deployed().then(function(instance) {instance.getAllAdmins().then(function(tx) {console.log(tx)});});

Reviewer: ```suggestion
Admin.deployed().then(function(instance) {instance.getAllAdmins().then(function(tx) {console.log(tx)});});
```

Author: Admin.deployed().then(function(instance) {instance.getAdmins().then(function(tx) {console.log(tx)});});

Author: Fixed

description: Onchain Permissioning
<!--- END of page meta data -->

# Onchain Permissioning 

Onchain permissioning uses smart contracts to store and maintain the node whitelist. Using onchain permissioning
enables all nodes to read the node whitelist from one location. 
                                                                       
!!! note
    Pantheon supports onchain permissioning for nodes. Onchain permissioning for accounts will be available 
    in a future Pantheon release. 

We have provided a set of smart contracts that interface with onchain permissioning in the 
[PegaSysEng/permissioning-smart-contracts](https://github.com/PegaSysEng/permissioning-smart-contracts) repository. 
We have also provided a management interface to interact with the contracts. 

## Provided Contracts

The following smart contracts are provided in the [PegaSysEng/permissioning-smart-contracts](https://github.com/PegaSysEng/permissioning-smart-contracts) repository: 

* Ingress - a simple contract functioning as a gateway to the Admin and Rules contracts. The Ingress contract is deployed 
to a static address. 

> * Rules - stores the node whitelist. 

Author: Fixed

> * Admin - stores the list of admin accounts.

Author: Fixed

## Pre-requisites 

> For nodes that are going to maintain the nodes whitelist or the list of admin accounts: 

Reviewer 2: Not sure if it's mentioned elsewhere - there's a few implicit configuration prerequisites: JSON-RPC enabled + appropriate methods etc.

Author: Following this up separately and will cover in a new PR if needed
